rust/lance-index/src/vector/graph/builder.rs

```diff
@@ -1,167 +1,161 @@
 // SPDX-License-Identifier: Apache-2.0
 // SPDX-FileCopyrightText: Copyright The Lance Authors
 
-use std::collections::{BinaryHeap, HashMap};
-use std::sync::Arc;
-
-use lance_core::{Error, Result};
-use snafu::{location, Location};
+use std::collections::BinaryHeap;
 
+use super::OrderedFloat;
 use super::OrderedNode;
-use super::{memory::InMemoryVectorStorage, Graph, GraphNode, OrderedFloat};
-use crate::vector::graph::storage::VectorStorage;
-use crate::vector::hnsw::select_neighbors_heuristic;
 
 /// GraphNode during build.
 #[derive(Debug, Clone)]
 pub struct GraphBuilderNode {
     /// Node ID
     pub(crate) id: u32,
 
-    /// Neighbors, sorted by the distance.
-    pub(crate) neighbors: BinaryHeap<OrderedNode>,
+    /// neighbors of each level of the node.
+    pub(crate) level_neighbors: Vec<BinaryHeap<OrderedNode>>,
 }
 
 impl GraphBuilderNode {
-    fn new(id: u32) -> Self {
+    pub(crate) fn new(id: u32, max_level: usize) -> Self {
         Self {
             id,
-            neighbors: BinaryHeap::new(),
+            level_neighbors: vec![BinaryHeap::new(); max_level],
         }
     }
 
-    fn add_neighbor(&mut self, distance: OrderedFloat, id: u32) {
-        self.neighbors.push(OrderedNode { dist: distance, id });
+    pub(crate) fn add_neighbor(&mut self, v: u32, dist: OrderedFloat, level: u16) {
+        self.level_neighbors[level as usize].push(OrderedNode { dist, id: v });
     }
 }
 
-impl From<&GraphBuilderNode> for GraphNode<u32> {
-    fn from(node: &GraphBuilderNode) -> Self {
-        let neighbors = node
-            .neighbors
-            .clone()
-            .into_sorted_vec()
-            .into_iter()
-            .map(|n| n.id)
-            .collect::<Vec<_>>();
-        Self {
-            id: node.id,
-            neighbors,
-        }
-    }
-}
+// impl From<&GraphBuilderNode> for GraphNode<u32> {
```

Comment:
should we just delete this?

Reply:
So here you changed `Levels<Node>` to `Node<Levels>`?
Curious what is the perf benefit of that.

But lets delete the commented out code nonetheless.